### ISSUE TITLE
Prepend method while being in top adds too much elements in the DOM

### ISSUE THREAD

**Issue by StriderRanger** [dhilt/ngx-ui-scroll#93]
When inspecting the DOM while working with [prepend method](https://dhilt.github.io/ngx-ui-scroll/#/adapter#append-prepend-sync), I noticed that prepending X numbers of items while **being in the top** of the list, injects X additional elements in the DOM even if this number exceeds by far the [bufferSize](https://dhilt.github.io/ngx-ui-scroll/#/settings#buffer-size) defined for the list.

I noticed also that this number of elements is kept in the dom as long as we are scrolling up or down the list while visualizing the new added items. The number of elements present in the dom is reduced to the bufferSize only when we are scrolling the old items.

Is there a way to keep always an approximate number to the bufferSize in this particular case?

Thanks

**Comment by dhilt** [comment]
@StriderRanger Thanks for the issue! Basically, it is not supposed that you can manage a number of DOM elements directly. The `bufferSize` setting defines a minimal number of items the UIScroll might retrieve per single `Datasource.get` call. So it seems impossible to have `bufferSize` elements in DOM no matter what the value of `bufferSize` is. 

I would say, if you want to have as minimal number of DOM elements as possible, you should set `bufferSize` and `padding` settings to their minimal values (1 and 0.01 respectively); but this will lead to maximum number of `Datasource.get` calls during the UIScroll work. The [settings page](https://dhilt.github.io/ngx-ui-scroll/#/settings) of the demo app tells the full story of the relation between DOM elements and the UIScroll settings.

At last, I didn't see any issues with `Adapter.prepend` and number of new DOM elements. [Adapter append/prepend sync sample](https://dhilt.github.io/ngx-ui-scroll/#/adapter#append-prepend-sync) demonstrates the consistency; if I set the input value to 10 and press append, I will get  exactly 10 new items rendered in 10 new DOM elements. If you have some specific issue with `Adapter.prepend` method, you need to give me a repro (you may fork [this project](https://stackblitz.com/edit/angular-ngx-ui-scroll-1-1-0-beta)) so I can to investigate it.

Hope this helps!

**Comment by StriderRanger** [comment]
Thank you for your reply!
The problem I am facing is not related to the bufferSize exact value as much as it is related to the 'infinite' adding of elements to the ui-scroll. This problem occurs as long as I call the prepend method while scrolling up the list.

To make things more clear, and as requested, I created [this example](https://stackblitz.com/edit/ngx-ui-scroll-prepend), and the results of my test are explained in this gif below:

![Prepend-Scroll-up](https://user-images.githubusercontent.com/15088185/63045761-8e2fb400-bec0-11e9-8091-d15713f49234.gif)

Thanks!

**Comment by dhilt** [comment]
@StriderRanger Now I see... the demo is very helpful. The thing is the UIScroll runs clipping process only after fetching new items via `Datasource.get`. I hear your concerns, let me think a bit what could be suggested to resolve them.

**Comment by StriderRanger** [comment]
Sounds good :) thank you very much for your support!

**Comment by dhilt** [comment]
It was decided to go with Adapter.clip() method that will provide removing out-of-viewport items by demand. The work had been started, the PR is the https://github.com/dhilt/ngx-ui-scroll/pull/97.

**Comment by StriderRanger** [comment]
That's great! I will apply the `Adapter.clip()` method once it is available and its development is complete. It will surely help.
Thanks :)

**Comment by dhilt** [comment]
@StriderRanger Big part of the work is done, there are some todos in PR #97, but 80% of required logic is implemented. It would be great if you could try it locally and give some feedback. You need to 
 - clone the repo
 - checkout "user-clip" branch
 - build the distributive via `npm run build`
 - place the distributive into the "node_modules" folder (replace version from the npm)

The usage is just to call `datasource.adapter.clip()` in appropriate moment. Do not call it if the UIScroll is busy, wait for it's rest. The `adapter.isLoading$` subscription can help.

**Comment by StriderRanger** [comment]
I cloned the repo and then checked out the "user-clip" branch, but by running `npm run build` (after `npm install`) I have this error:

> Error: Can't define package version
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! ngx-ui-scroll-app@0.0.0 build: `node build.js`
> npm ERR! Exit status 1

**Comment by dhilt** [comment]
@StriderRanger You do it on Win, don't you? let me think, how to update the build process to fix it...

**Comment by dhilt** [comment]
Created bug https://github.com/dhilt/ngx-ui-scroll/issues/99

**Comment by dhilt** [comment]
@StriderRanger The build script had been fixed, it should work now on Win OS. Also, the development of `Adapter.clip` feature is 95 per cent done.

**Comment by StriderRanger** [comment]
@dhilt Yes I am on windows, thanks for the fix! 
It's good to know that the `Adapter.clip` is almost complete. I will update and launch a new build, and let you know of the results of my tests.

**Comment by StriderRanger** [comment]
The `Adapter.clip()` method did work. Here is the procedure that I followed:
Trying first to subscribe to the `Adapter.isLoading$` in the constructor, and then call the clip inside the subscription didn't work: The list is clipped at the first load and there was no scroll available.
What I did instead is calling the `Adapter.clip()` before the prepend: The idea was to clip the list from the ancient items before prepending a new ones. In my [Stackblitz example](https://stackblitz.com/edit/ngx-ui-scroll-prepend) I added in a comment the line where I call the `Adapter.clip()` (line 62)
Let me know if you need to know anything else about this test.
Thanks!

**Comment by dhilt** [comment]
@StriderRanger My vision of the isLoading-deferred method call is following:
```
    const { adapter } = this.datasource;
    if (!adapter.isLoading) {
      cb();
      return;
    }
    const isLoadingSub = adapter.isLoading$.subscribe(isLoading => {
      if (!isLoading) {
        isLoadingSub.unsubscribe();
        cb();
      }
    });
```
This is one-time self-destructing subscription allowing to run necessary piece of logic (cb) right after the uiScroll Workflow stops.

I published the "@next" version of ngx-ui-scroll to npm and updated your demo:
https://stackblitz.com/edit/ngx-ui-scroll-prepend-1-1-1-experimental. For me it works properly.

**Comment by StriderRanger** [comment]
@dhilt I understand more clearly now how the `adapter.isLoading$` subscription should be used. The one-time self-destructing subscription is a nice and clean solution, and it made the scroll-top go as expected. 
I appreciate your support and your prompt action! Thank you :)

**Comment by dhilt** [comment]
v1.1.1 ("Adapter clip") has been released
Thanks for the issue!
